electricitymap/contrib/lib/models/events.py

```diff
@@ -0,0 +1,357 @@
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta, timezone
+from enum import Enum
+from logging import Logger
+from typing import Any, Dict, Optional
+
+from pydantic import BaseModel, PrivateAttr, validator
+
+from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG, ZoneKey
+from electricitymap.contrib.lib.models.constants import VALID_CURRENCIES
+
+LOWER_DATETIME_BOUND = datetime(2000, 1, 1, tzinfo=timezone.utc)
+
+
+class Mix(BaseModel, ABC):
+    def set_value(self, mode: str, value: float) -> None:
+        """
+        Sets the value of a production mode.
+        This can be used if the Production has been initialized empty
+        and is being filled in a loop.
+        """
+        self.__setattr__(mode, value)
+
+
+class ProductionMix(Mix):
+    """
+    Contains the production mix for a zone at a given time.
+    All values should be positives, otherwise they will be set to None
+    and a warning will be logged.
+    All values are in MW.
+    """
+
+    # We use a private attribute to keep track of the modes that have been set to None.
+    _corrected_negative_values: set = PrivateAttr(set())
+    biomass: Optional[float] = None
+    coal: Optional[float] = None
+    gas: Optional[float] = None
+    geothermal: Optional[float] = None
+    hydro: Optional[float] = None
+    nuclear: Optional[float] = None
+    oil: Optional[float] = None
+    solar: Optional[float] = None
+    unknown: Optional[float] = None
+    wind: Optional[float] = None
+
+    def __init__(self, **data: Any):
+        """Overriding the constructor to check for negative values and set them to None."""
+        super().__init__(**data)
+        for attr, value in data.items():
+            if value is not None and value < 0:
+                self._corrected_negative_values.add(attr)
+                self.__setattr__(attr, None)
+
+    def set_value(self, mode: str, value: float) -> None:
```

Comment:
I think it might be better to do something like:

(giving a whole snippet to play around)

```
from typing import Optional

from pydantic import BaseModel, PrivateAttr


class ProductionMix(BaseModel):
    """
    Contains the production mix for a zone at a given time.
    All values should be positives, otherwise they will be set to None
    and a warning will be logged.
    All values are in MW.
    """

    # We use a private attribute to keep track of the modes that have been set to None.
    _corrected_negative_values: set = PrivateAttr(set())
    biomass: Optional[float] = None
    coal: Optional[float] = None
    gas: Optional[float] = None
    geothermal: Optional[float] = None
    hydro: Optional[float] = None
    nuclear: Optional[float] = None
    oil: Optional[float] = None
    solar: Optional[float] = None
    unknown: Optional[float] = None
    wind: Optional[float] = None

    def __setattr__(self, name, value):
        if name in ["biomass"]:
            if value < 0:
                self._corrected_negative_values.add(name)
                print(
                    f"Negative value for {name} production"
                    f"Setting to None."
                )
                value = None
            super().__setattr__(name, value)


if __name__ == "__main__":
    pt = ProductionMix()
    breakpoint()
```

This way there is no way for a contributor to unknowingly circumvet this logic by doing `mix.biomass = -1`, and it makes the setting less verbose

Reply:
(still need to also do the check at init)

Reply:
What about this @unitrium ?

Reply:
I oversaw this comment since it was made separately. I will take a look.

Reply:
We discussed this with @mathilde-daugy  and thought that it might be overwhelming for beginner contributors to use the __setattr__ since some might not know this default method.
However I agree that it introduces the risk of a contributor using it and by passing our verification.

Reply:
So I would rather settle for this.

Reply:
Added this and kept the set_mode for consistency with storage mix.